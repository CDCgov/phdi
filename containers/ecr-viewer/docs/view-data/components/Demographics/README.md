[**ecr-viewer**](../../../README.md) â€¢ **Docs**

***

[ecr-viewer](../../../README.md) / view-data/components/Demographics

# view-data/components/Demographics

## Index

### Functions

- [default](functions/default.md)
