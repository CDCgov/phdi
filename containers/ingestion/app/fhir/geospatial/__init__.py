from app.fhir.geospatial.census import CensusFhirGeocodeClient
from app.fhir.geospatial.core import BaseFhirGeocodeClient
from app.fhir.geospatial.smarty import SmartyFhirGeocodeClient

__all__ = (
    "BaseFhirGeocodeClient",
    "SmartyFhirGeocodeClient",
    "CensusFhirGeocodeClient",
)
