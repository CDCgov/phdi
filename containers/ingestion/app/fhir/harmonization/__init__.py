from app.fhir.harmonization.standardization import double_metaphone_bundle
from app.fhir.harmonization.standardization import double_metaphone_patient
from app.fhir.harmonization.standardization import standardize_dob
from app.fhir.harmonization.standardization import standardize_names
from app.fhir.harmonization.standardization import standardize_phones

__all__ = (
    "double_metaphone_bundle",
    "double_metaphone_patient",
    "standardize_names",
    "standardize_phones",
    "standardize_dob",
)
