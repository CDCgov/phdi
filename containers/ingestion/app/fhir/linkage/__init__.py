from app.fhir.linkage.link import add_patient_identifier
from app.fhir.linkage.link import add_patient_identifier_in_bundle

__all__ = ["add_patient_identifier_in_bundle", "add_patient_identifier"]
