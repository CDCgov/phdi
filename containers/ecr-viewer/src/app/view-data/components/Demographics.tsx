import { evaluate } from "fhirpath";
import { Bundle } from "fhir/r4";
import { PathMappings } from "@/app/utils";

interface DemographicsProps {
    fhirPathMappings: PathMappings
    fhirBundle: Bundle | undefined
}



const patientName = (fhirBundle: Bundle | undefined, fhirPathMappings: PathMappings) => {
    const givenNames = evaluate(fhirBundle, fhirPathMappings.patientGivenName).join(" ");
    const familyName = evaluate(fhirBundle, fhirPathMappings.patientFamilyName);

    return `${givenNames} ${familyName}`;
}


const Demographics = (
    { fhirPathMappings, fhirBundle }: DemographicsProps
) => {
    return (
        <div>
            <div
                className="padding-bottom-3"
                aria-labelledby="summary-box-key-information"
            >
                <div className="usa-summary-box__body">
                    <h3
                        className="usa-summary-box__heading padding-y-105"
                        id="summary-box-key-information"
                    >
                        Demographics
                    </h3>
                    <div className="usa-summary-box__text">
                        <div className="grid-row">
                            <div className="data-title"><h4>Patient Name</h4></div>
                            <div className="grid-col-auto">
                                {patientName(fhirBundle, fhirPathMappings)}
                            </div>
                        </div>
                        <div className={"section__line_gray"} />
                        <div className="grid-row">
                            <div className="data-title"><h4>Patient ID</h4></div>
                            <div className="grid-col-auto">
                                {evaluate(fhirBundle, fhirPathMappings.patientId)}
                            </div>
                        </div>
                        <div className={"section__line_gray"} />
                        <div className="grid-row">
                            <div className="data-title"><h4>DOB</h4></div>
                            <div className="grid-col-auto">
                                {evaluate(fhirBundle, fhirPathMappings.patientDOB)}
                            </div>
                        </div>
                        <div className={"section__line_gray"} />
                        <div className="grid-row">
                            <div className="data-title"><h4>Sex</h4></div>
                            <div className="grid-col-auto">
                                {evaluate(fhirBundle, fhirPathMappings.patientGender)}
                            </div>
                        </div>
                        <div className={"section__line_gray"} />
                        <div className="grid-row">
                            <div className="data-title"><h4>Race</h4></div>
                            <div className="grid-col-auto">
                                {evaluate(fhirBundle, fhirPathMappings.patientRace)}
                            </div>
                        </div>
                        <div className={"section__line_gray"} />
                        <div className="grid-row">
                            <div className="data-title"><h4>Ethnicity</h4></div>
                            <div className="grid-col-auto">
                                {evaluate(fhirBundle, fhirPathMappings.patientEthnicity)}
                            </div>
                        </div>
                        <div className={"section__line_gray"} />
                        <div className="grid-row">
                            <div className="data-title"><h4>Preferred Language</h4></div>
                            <div className="grid-col-auto">
                                {evaluate(fhirBundle, fhirPathMappings.patientLanguage)}
                            </div>
                        </div>
                        <div className={"section__line_gray"} />
                        <div className="grid-row">
                            <div className="data-title"><h4>Patient Address</h4></div>
                            <div className="grid-col-auto text-pre-line">
                                {evaluate(fhirBundle, fhirPathMappings.patientStreetAddress).join("\n")} {'\n'}
                                {evaluate(fhirBundle, fhirPathMappings.patientCity)}, {evaluate(fhirBundle, fhirPathMappings.patientState)} {'\n'}
                                {evaluate(fhirBundle, fhirPathMappings.patientZipCode)}, USA
                            </div>
                        </div>
                        <div className={"section__line_gray"} />
                        <div className="grid-row">
                            <div className="data-title"><h4>Contact</h4></div>
                            <div className="grid-col-auto text-pre-line">
                                {evaluate(fhirBundle, fhirPathMappings.patientPhoneNumbers).map(phoneNumber => {
                                    return `tel: (${phoneNumber.use}) ${phoneNumber.value}`
                                }).join("\n")}
                                {'\n'}
                                {evaluate(fhirBundle, fhirPathMappings.patientEmails).map(email => {
                                    return `email: ${email.value}`
                                }).join("\n")}
                            </div>
                        </div>
                        <div className={"section__line_gray"} />
                    </div>
                </div>
            </div>
        </div>);
};

export default Demographics;