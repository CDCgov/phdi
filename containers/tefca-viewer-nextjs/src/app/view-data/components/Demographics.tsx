import { Patient, HumanName, Address, ContactPoint, Identifier } from 'fhir/r4';
import { DisplayData } from '@/app/utils';
import { DataDisplay } from '@/app/utils';
import * as dateFns from "date-fns";
import {evaluate} from "fhirpath";

export interface DemographicsProps {
    patient: Patient;
}

function formatDemographics(patient: Patient) {

    const demographicData: DisplayData[] = [
        {
          title: "Patient Name",
          value: formatName(patient.name ?? []),
        },
        { 
          title: "DOB",
          value: patient.birthDate ?? "",
        },
        {
          title: "Current Age",
          value: calculatePatientAge(patient)?.toString(),
        },
        { 
          title: "Sex",
          value: patient.gender ?? "",
        },
        { 
          title: "Race",
          value: evaluate(patient, "Patient.extension.where(url = 'http://hl7.org/fhir/us/core/StructureDefinition/us-core-ethnicity').extension.first().valueCoding.display")[0],
        },
        {
          title: "Ethnicity",
          value: evaluate(patient, "Patient.extension.where(url = 'http://hl7.org/fhir/us/core/StructureDefinition/us-core-race').extension.first().valueCoding.display")[0],
        },
        {
          title: "Tribal Affiliation",
          value: evaluate(patient, "Patient.extension.where(url='http: //hl7.org/fhir/us/ecr/StructureDefinition/us-ph-tribal-affiliation-extension').extension.where(url='TribeName').value.display")[0],
        },
        {
          title: "Preferred Language",
          value: evaluate(patient, "Patient.communication.first().language.coding.first().display")[0],
        },
        {
          title: "Address",
          value: formatAddress(patient.address ?? []),
        },
        {
          title: "Contact",
          value: formatContact(patient.telecom ?? []),
        },
        {
          title: "Patient Identifiers",
          value: formatIdentifier(patient.identifier ?? []),
        },
      ];

    return demographicData;
}

function formatName(names: HumanName[]) {
    let name = "";
    if (names.length > 0) {
        name = names[0].given?.join(" ") + " " + names[0].family;
    }
    return name;
}

function formatAddress(address: Address[]) {
    let formattedAddress = "";
    if (address.length > 0) {
        formattedAddress = address[0]?.line?.join("\n") + "\n" + address[0].city + ", " + address[0].state + " " + address[0].postalCode;
    }
    return formattedAddress;
}

function formatContact(contacts: ContactPoint[]) {
    return (
        contacts.map((contact) => {
            if (contact.system === "phone") {
                return `${contact.use}: ${contact.value}`;
            }
            else if (contact.system === "email") {
                return contact.value;
            }
        }
        ).join("\n")
    )
}

function formatIdentifier(identifier: Identifier[]) {
    return (
        identifier.map((id) => {
            let idType = id.type?.coding?.[0].display ?? "";
            if (idType === "") {
                idType = id.type?.text ?? "";
            }

            return `${idType}: ${id.value}`;
        }
        ).join("\n")
    )
}

export const calculatePatientAge = (patient: Patient) => {

    if (patient.birthDate) {
      const patientDOB = new Date(patient.birthDate);
      const today = new Date();
      return dateFns.differenceInYears(today, patientDOB);
    }
  };

export default function Demographics(props: DemographicsProps) {

    const demographicData = formatDemographics(props.patient);

    return (
        <div>
            {demographicData.map((item, index) => (
                <DataDisplay item={item} key={index} />
            ))}
        </div>
    );

}