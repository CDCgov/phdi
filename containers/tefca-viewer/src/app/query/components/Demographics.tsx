import { Patient, HumanName, Address, ContactPoint, Identifier } from "fhir/r4";
import { DisplayData } from "@/app/utils";
import { DataDisplay } from "@/app/utils";
import * as dateFns from "date-fns";
import { evaluate } from "fhirpath";

/**
 * Displays the demographic information of a patient.
 * @param {Patient} patient - The patient to display demographic information for.
 * @returns {React.FC} The Demographics component.
 */
export interface DemographicsProps {
  patient: Patient;
}

/**
 * Displays the demographic information of a patient.
 * @param props - The props for the Demographics component.
 * @param props.patient - The patient resource to display demographic information for.
 * @returns The Demographics component.
 */
const Demographics: React.FC<DemographicsProps> = ({ patient }) => {
  const demographicData = formatDemographics(patient);

  return (
    <div>
      {demographicData.map((item, index) => (
        <DataDisplay item={item} key={index} />
      ))}
    </div>
  );
};

export default Demographics;

/**
 * Formats the demographic information of a patient.
 * @param patient - The patient to format demographic information for.
 * @returns The formatted demographic information as an array of DisplayData objects.
 */
function formatDemographics(patient: Patient): DisplayData[] {
  const demographicData: DisplayData[] = [
    {
      title: "Patient Name",
      value: formatName(patient.name ?? []),
    },
    {
      title: "DOB",
      value: patient.birthDate ?? "",
    },
    {
      title: "Current Age",
      value: calculatePatientAge(patient)?.toString(),
    },
    {
      title: "Sex",
      value: patient.gender ?? "",
    },
    {
      title: "Race",
      value: evaluate(
        patient,
        "Patient.extension.where(url = 'http://hl7.org/fhir/us/core/StructureDefinition/us-core-ethnicity').extension.first().valueCoding.display",
      )[0],
    },
    {
      title: "Ethnicity",
      value: evaluate(
        patient,
        "Patient.extension.where(url = 'http://hl7.org/fhir/us/core/StructureDefinition/us-core-race').extension.first().valueCoding.display",
      )[0],
    },
    {
      title: "Tribal Affiliation",
      value: evaluate(
        patient,
        "Patient.extension.where(url='http: //hl7.org/fhir/us/ecr/StructureDefinition/us-ph-tribal-affiliation-extension').extension.where(url='TribeName').value.display",
      )[0],
    },
    {
      title: "Preferred Language",
      value: evaluate(
        patient,
        "Patient.communication.first().language.coding.first().display",
      )[0],
    },
    {
      title: "Address",
      value: formatAddress(patient.address ?? []),
    },
    {
      title: "Contact",
      value: formatContact(patient.telecom ?? []),
    },
    {
      title: "Patient Identifiers",
      value: formatIdentifier(patient.identifier ?? []),
    },
  ];

  return demographicData;
}

/**
 * Formats the name of a FHIR HumanName object.
 * @param names - The HumanName object to format.
 * @returns The formatted name.
 */
function formatName(names: HumanName[]): string {
  let name = "";
  if (names.length > 0) {
    name = names[0].given?.join(" ") + " " + names[0].family;
  }
  return name;
}

/**
 * Formats the address of a FHIR Address object.
 * @param address - The Address object to format.
 * @returns The formatted address.
 */
function formatAddress(address: Address[]): string {
  let formattedAddress = "";
  if (address.length > 0) {
    formattedAddress =
      address[0]?.line?.join("\n") +
      "\n" +
      address[0].city +
      ", " +
      address[0].state +
      " " +
      address[0].postalCode;
  }
  return formattedAddress;
}

/**
 * Formats the contact information of a FHIR ContactPoint object.
 * @param contacts - The ContactPoint object to format.
 * @returns - The formatted contact information.
 */
function formatContact(contacts: ContactPoint[]): string {
  return contacts
    .map((contact) => {
      if (contact.system === "phone") {
        return `${contact.use}: ${contact.value}`;
      } else if (contact.system === "email") {
        return contact.value;
      }
    })
    .join("\n");
}

/**
 * Formats the identifiers of a FHIR Identifier object.
 * @param identifier - The Identifier object to format.
 * @returns The formatted identifiers.
 */
function formatIdentifier(identifier: Identifier[]): string {
  return identifier
    .map((id) => {
      let idType = id.type?.coding?.[0].display ?? "";
      if (idType === "") {
        idType = id.type?.text ?? "";
      }

      return `${idType}: ${id.value}`;
    })
    .join("\n");
}

/**
 * Calculates the age of a patient based on their birth date.
 * @param patient - The patient to calculate the age for.
 * @returns The age of the patient.
 */
export function calculatePatientAge(patient: Patient): number | undefined {
  if (patient.birthDate) {
    const patientDOB = new Date(patient.birthDate);
    const today = new Date();
    return dateFns.differenceInYears(today, patientDOB);
  }
}
