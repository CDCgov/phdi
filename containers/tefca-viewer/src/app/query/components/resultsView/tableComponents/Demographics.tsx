import { Patient } from "fhir/r4";
import { DataDisplayInfo } from "@/app/utils";
import { DataDisplay } from "@/app/utils";
import * as dateFns from "date-fns";
import { evaluate } from "fhirpath";

import {
  formatName,
  formatAddress,
  formatContact,
  formatIdentifier,
  formatDate,
} from "../../../../format-service";

/**
 * Displays the demographic information of a patient.
 * @param {Patient} patient - The patient to display demographic information for.
 * @returns {React.FC} The Demographics component.
 */
export interface DemographicsProps {
  patient: Patient;
}

/**
 * Displays the demographic information of a patient.
 * @param props - The props for the Demographics component.
 * @param props.patient - The patient resource to display demographic information for.
 * @returns The Demographics component.
 */
const Demographics: React.FC<DemographicsProps> = ({ patient }) => {
  const demographicData = formatDemographics(patient);

  return (
    <div>
      {demographicData.map((item) => (
        <DataDisplay item={item} key={item.title} />
      ))}
    </div>
  );
};

export default Demographics;

/**
 * Formats the demographic information of a patient.
 * @param patient - The patient to format demographic information for.
 * @returns The formatted demographic information as an array of DisplayData objects.
 */
function formatDemographics(patient: Patient): DataDisplayInfo[] {
  const demographicData: DataDisplayInfo[] = [
    {
      title: "Patient Name",
      value: formatName(patient.name ?? []),
    },
    {
      title: "DOB",
      value: formatDate(patient.birthDate),
    },
    {
      title: "Current Age",
      value: calculatePatientAge(patient)?.toString(),
    },
    {
      title: "Sex",
      value: patient.gender ?? "",
    },
    {
      title: "Race",
      value: evaluate(
        patient,
        "Patient.extension.where(url = 'http://hl7.org/fhir/us/core/StructureDefinition/us-core-race').extension.first().valueCoding.display",
      )[0],
    },
    {
      title: "Ethnicity",
      value: evaluate(
        patient,
        "Patient.extension.where(url = 'http://hl7.org/fhir/us/core/StructureDefinition/us-core-ethnicity').extension.first().valueCoding.display",
      )[0],
    },
    {
      title: "Tribal Affiliation",
      value: evaluate(
        patient,
        "Patient.extension.where(url='http: //hl7.org/fhir/us/ecr/StructureDefinition/us-ph-tribal-affiliation-extension').extension.where(url='TribeName').value.display",
      )[0],
    },
    {
      title: "Preferred Language",
      value: evaluate(
        patient,
        "Patient.communication.first().language.coding.first().display",
      )[0],
    },
    {
      title: "Address",
      value: formatAddress(patient.address ?? []),
    },
    {
      title: "Contact",
      value: formatContact(patient.telecom ?? []),
    },
    {
      title: "Patient Identifiers",
      value: formatIdentifier(patient.identifier ?? []),
    },
  ];

  return demographicData;
}

/**
 * Calculates the age of a patient based on their birth date.
 * @param patient - The patient to calculate the age for.
 * @returns The age of the patient.
 */
export function calculatePatientAge(patient: Patient): number | undefined {
  if (patient.birthDate) {
    const patientDOB = new Date(patient.birthDate);
    const today = new Date();
    return dateFns.differenceInYears(today, patientDOB);
  }
}
