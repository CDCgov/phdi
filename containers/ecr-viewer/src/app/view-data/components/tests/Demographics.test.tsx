import { render } from "@testing-library/react";
import fs from "fs";
import { Bundle } from "fhir/r4";
import YAML from "yaml";
import { axe } from "jest-axe";
import Demographics from "@/app/view-data/components/Demographics";

describe("Demographics", () => {
  let container: HTMLElement;
  beforeAll(() => {
    const fhirPathFile = fs
      .readFileSync("./src/app/api/fhir-data/fhirPath.yml", "utf8")
      .toString();
    const fhirPathMappings = YAML.parse(fhirPathFile);
    const fhirBundle: Bundle = JSON.parse(
      fs
        .readFileSync(
          "./seed-scripts/fhir_data/1dd10047-2207-4eac-a993-0f706c88be5d.json",
          "utf8",
        )
        .toString(),
    );

    container = render(
      <Demographics
        fhirPathMappings={fhirPathMappings}
        fhirBundle={fhirBundle}
      />,
    ).container;
  });
  it("should match snapshot", () => {
    expect(container).toMatchSnapshot();
  });
  it("should pass accessibility test", async () => {
    expect(await axe(container)).toHaveNoViolations();
  });
});
