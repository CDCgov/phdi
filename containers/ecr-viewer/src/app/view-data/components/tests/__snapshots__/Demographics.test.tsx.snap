// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Demographics should match snapshot 1`] = `
<div>
  <div>
    <div
      aria-labelledby="summary-box-key-information"
      class="padding-bottom-3"
    >
      <div
        class="usa-summary-box__body"
      >
        <h3
          class="usa-summary-box__heading padding-y-105"
          id="summary-box-key-information"
        >
          Demographics
        </h3>
        <div
          class="usa-summary-box__text"
        >
          <div
            class="grid-row"
          >
            <div
              class="data-title"
            >
              <h4>
                Patient Name
              </h4>
            </div>
            <div
              class="grid-col-auto"
            >
              ABEL CASTILLO
            </div>
          </div>
          <div
            class="section__line_gray"
          />
          <div
            class="grid-row"
          >
            <div
              class="data-title"
            >
              <h4>
                DOB
              </h4>
            </div>
            <div
              class="grid-col-auto"
            >
              2015-04-15
            </div>
          </div>
          <div
            class="section__line_gray"
          />
          <div
            class="grid-row"
          >
            <div
              class="data-title"
            >
              <h4>
                Sex
              </h4>
            </div>
            <div
              class="grid-col-auto"
            >
              male
            </div>
          </div>
          <div
            class="section__line_gray"
          />
          <div
            class="grid-row"
          >
            <div
              class="data-title"
            >
              <h4>
                Race
              </h4>
            </div>
            <div
              class="grid-col-auto"
            >
              Black or African American
            </div>
          </div>
          <div
            class="section__line_gray"
          />
          <div
            class="grid-row"
          >
            <div
              class="data-title"
            >
              <h4>
                Ethnicity
              </h4>
            </div>
            <div
              class="grid-col-auto"
            >
              Hispanic or Latino
            </div>
          </div>
          <div
            class="section__line_gray"
          />
          <div
            class="grid-row"
          >
            <div
              class="data-title"
            >
              <h4>
                Tribal Affiliation
              </h4>
            </div>
            <div
              class="grid-col-auto"
            />
          </div>
          <div
            class="section__line_gray"
          />
          <div
            class="grid-row"
          >
            <div
              class="data-title"
            >
              <h4>
                Preferred Language
              </h4>
            </div>
            <div
              class="grid-col-auto"
            >
              English
            </div>
          </div>
          <div
            class="section__line_gray"
          />
          <div
            class="grid-row"
          >
            <div
              class="data-title"
            >
              <h4>
                Patient Address
              </h4>
            </div>
            <div
              class="grid-col-auto text-pre-line"
            >
              1050 CARPENTER ST
    EDWARDS, CA
        93523-2800, US
            </div>
          </div>
          <div
            class="section__line_gray"
          />
          <div
            class="grid-row"
          >
            <div
              class="data-title"
            >
              <h4>
                Contact
              </h4>
            </div>
            <div
              class="grid-col-auto text-pre-line"
            >
              Home +18184195968
    MELLY.C.A.16@GMAIL.COM
            </div>
          </div>
          <div
            class="section__line_gray"
          />
          <div
            class="grid-row"
          >
            <div
              class="data-title"
            >
              <h4>
                Emergency Contact
              </h4>
            </div>
            <div
              class="grid-col-auto text-pre-line"
            >
              N/A
            </div>
          </div>
          <div
            class="section__line_gray"
          />
          <div
            class="grid-row"
          >
            <div
              class="data-title"
            >
              <h4>
                Patient ID
              </h4>
            </div>
            <div
              class="grid-col-auto"
            >
              6b6b3c4c-4884-4a96-b6ab-c46406839cea
            </div>
          </div>
          <div
            class="section__line_gray"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
